File: test/ts-tableau-loader/tests/load.test.ts
// Load block: hub loading, filter, binary-format loading and patch loading.
// Mirrors the same scenarios in:
//   - Go:  test/go-tableau-loader/load_test.go
//   - C#:  test/csharp-tableau-loader/tests/LoadTests.cs
//   - C++: test/cpp-tableau-loader/tests/load_test.cpp
import { equals } from "@bufbuild/protobuf";

import { Hub, Registry } from "../tableau/hub.pc.js";
import { Format, LoadMode } from "../tableau/load.pc.js";
import {
  PatchMergeConf,
  PatchReplaceConf,
  RecursivePatchConf,
} from "../tableau/patch_conf.pc.js";
import { HeroConf } from "../tableau/hero_conf.pc.js";
import * as protoconf from "../tableau/barrel/protoconf.pc.js";

import { CustomItemConf, CustomItemConfName } from "../custom/custom_item_conf.js";

import {
  assert,
  check,
  confDir,
  binDir,
  patchConf,
  patchConf2,
  patchResult,
  testdata,
  join,
  prepareHub,
} from "./harness.js";

// Register the hand-written CustomItemConf before any hub is built, so the
// shared hub (and any hub loaded afterwards) includes it. Mirrors Go's init()
// Register, C#'s HubFixture registration and C++'s InitCustomMessager.
Registry.register(CustomItemConf);

export function run(): void {
  // ---- Load ----

  // Hub loads all messagers from a JSON directory.
  check("hub.load(conf, JSON)", () => {
    const hub = prepareHub();
    assert.ok(hub.getItemConf());
    assert.ok(hub.getActivityConf());
  });

  // ---- CustomConf ----

  // CustomItemConf is a hand-written messager: it has no file of its own and
  // resolves a "special" item from ItemConf via processAfterLoadAll. Mirrors
  // Go's Test_CustomItemConf and C#/C++'s
  // CustomItemConf_ProcessAfterLoadAll_ResolvesSpecialItem.
  check("CustomItemConf processAfterLoadAll", () => {
    const hub = prepareHub();
    const custom = hub.getMessager(CustomItemConfName) as CustomItemConf | undefined;
    assert.ok(custom);
    // ItemConf item 1 is "apple", so the resolved special item name is non-empty.
    assert.notEqual(custom!.getSpecialItemName(), "");
    assert.equal(custom!.getSpecialItemName(), "apple");
  });

  // Hub typed accessor for an unloaded-by-filter messager returns undefined.
  check("Hub filter", () => {
    const filtered = new Hub({ filter: (n) => n === "ItemConf" });
    filtered.load(confDir, Format.JSON, { ignoreUnknownFields: true });
    assert.ok(filtered.getItemConf());
    assert.equal(filtered.getActivityConf(), undefined);
  });

  // ---- Bin ----

  // Binary format load via a single messager.
  check("HeroConf BIN load", () => {
    const hero = new HeroConf();
    hero.load(binDir, Format.BIN);
    assert.equal(hero.name(), "HeroConf");
  });

  // BIN load from a non-existent directory throws (mirrors C#'s failure case).
  check("HeroConf BIN load failure", () => {
    const hero = new HeroConf();
    assert.throws(() => hero.load(join(testdata, "no-such-bin-dir"), Format.BIN));
  });

  // ---- Patch ----

  // Patch MERGE: main + two patch files merged in order.
  check("PatchMergeConf MERGE", () => {
    const pm = new PatchMergeConf();
    pm.load(confDir, Format.JSON, {
      ignoreUnknownFields: true,
      patchPaths: [
        join(patchConf, "PatchMergeConf.json"),
        join(patchConf2, "PatchMergeConf.json"),
      ],
    });
    // Scalar overridden by the first patch.
    assert.equal(pm.data().name, "orange");
    // itemMap key 999 contributed by the second patch (merge, not replace).
    assert.ok(pm.get1(999));
    assert.ok(pm.get1(1));
    // itemMap merges field-by-field, so the main file's key 2 survives.
    assert.ok(pm.get1(2));
    // replaceItemMap has the field-level PATCH_REPLACE option: the last patch
    // that carries it fully replaces the field. patchconf2 provides {1, 999},
    // so key 999 is present while the main file's key 2 is dropped (replace,
    // not merge). Mirrors Go/C#/C++'s ReplaceItemMap[999] assertion.
    assert.ok(pm.data().replaceItemMap[999]);
    assert.equal(pm.data().replaceItemMap[2], undefined);
  });

  // LoadMode.ONLY_MAIN ignores the patch files.
  check("PatchMergeConf ONLY_MAIN", () => {
    const pm = new PatchMergeConf();
    pm.load(confDir, Format.JSON, {
      ignoreUnknownFields: true,
      mode: LoadMode.ONLY_MAIN,
      patchPaths: [
        join(patchConf, "PatchMergeConf.json"),
        join(patchConf2, "PatchMergeConf.json"),
      ],
    });
    assert.equal(pm.data().name, "apple");
    assert.equal(pm.get1(999), undefined);
  });

  // PATCH_MERGE golden: conf + patchconf merged must equal the canonical
  // patchresult, checked as a whole (mirrors Go/C#'s proto.Equal comparison).
  check("RecursivePatchConf MERGE golden", () => {
    const got = new RecursivePatchConf();
    got.load(confDir, Format.JSON, {
      ignoreUnknownFields: true,
      patchDirs: [patchConf],
    });

    // The golden file is itself a RecursivePatchConf sheet (PATCH_MERGE), so load
    // it with no patch sources to get the main file content verbatim.
    const expected = new RecursivePatchConf();
    expected.load(patchResult, Format.JSON, { ignoreUnknownFields: true });

    assert.ok(
      equals(protoconf.RecursivePatchConfSchema, got.message(), expected.message()),
      "patched RecursivePatchConf does not match the golden patchresult",
    );
  });

  // PATCH_REPLACE: the sheet-level option replaces the whole message with the
  // last patch file, rather than merging field-by-field.
  check("PatchReplaceConf PATCH_REPLACE", () => {
    const pr = new PatchReplaceConf();
    pr.load(confDir, Format.JSON, {
      ignoreUnknownFields: true,
      patchDirs: [patchConf],
    });
    // Main file has name "apple" / priceList [10,100]; the patch fully replaces
    // it with name "orange" / priceList [20,200] (replace, not append).
    assert.equal(pr.data().name, "orange");
    assert.deepEqual(pr.data().priceList.map(Number), [20, 200]);
  });

  // LoadMode.ONLY_PATCH starts from an empty message and applies patches only.
  check("PatchMergeConf ONLY_PATCH", () => {
    const pm = new PatchMergeConf();
    pm.load(confDir, Format.JSON, {
      ignoreUnknownFields: true,
      mode: LoadMode.ONLY_PATCH,
      patchPaths: [
        join(patchConf, "PatchMergeConf.json"),
        join(patchConf2, "PatchMergeConf.json"),
      ],
    });
    // Name comes from the first patch (second patch carries no name field), and
    // both itemMap entries come purely from the patches, not the main file.
    assert.equal(pm.data().name, "orange");
    assert.ok(pm.get1(1));
    assert.ok(pm.get1(999));
  });
}

File: test/cpp-tableau-loader/tests/load_test.cpp
// Load block: hub loading, custom conf (ProcessAfterLoadAll), binary-format
// loading and patch loading. Mirrors the same scenarios in:
//   - Go:  test/go-tableau-loader/load_test.go
//   - C#:  test/csharp-tableau-loader/tests/LoadTests.cs
//   - TS:  test/ts-tableau-loader/tests/load.test.ts

#include <google/protobuf/util/message_differencer.h>
#include <gtest/gtest.h>

#include "hub/custom/item/custom_item_conf.h"
#include "protoconf/hero_conf.pc.h"
#include "protoconf/item_conf.pc.h"
#include "protoconf/patch_conf.pc.h"
#include "tests/hub_fixture.h"

namespace {

using ::google::protobuf::util::MessageDifferencer;

// ---- Load ----

TEST_F(HubFixture, Load_AllMessagers_Succeeds) {
  // SetUp() already asserts the hub loads; verify a few messagers are present.
  EXPECT_NE(Hub::Instance().Get<protoconf::ItemConfMgr>(), nullptr);
  EXPECT_NE(Hub::Instance().Get<tableau::ActivityConf>(), nullptr);
  EXPECT_NE(Hub::Instance().Get<tableau::FruitConf>(), nullptr);
}

// ---- CustomConf ----

TEST_F(HubFixture, CustomItemConf_ProcessAfterLoadAll_ResolvesSpecialItem) {
  auto custom = Hub::Instance().Get<CustomItemConf>();
  ASSERT_NE(custom, nullptr);
  EXPECT_FALSE(custom->GetSpecialItemName().empty());
}

// ---- Bin ----

TEST_F(HubFixture, HeroConf_LoadFromBin_Succeeds) {
  tableau::HeroConf hero_conf;
  bool ok = hero_conf.Load(test::TestPaths::Bin().string() + "/", tableau::Format::kBin);
  ASSERT_TRUE(ok) << "failed to load HeroConf.binpb: " << tableau::GetErrMsg();
}

TEST_F(HubFixture, HeroConf_LoadFromMissingDir_Fails) {
  tableau::HeroConf hero_conf;
  auto missing = (test::TestPaths::Testdata() / "notexist").string() + "/";
  EXPECT_FALSE(hero_conf.Load(missing, tableau::Format::kBin));
}

// ---- Patch ----
// PatchTest uses its own fixture: it only initializes the registry (InitOnce)
// and loads a fresh hub per test with patch options, mirroring the C#/Go patch
// tests that construct a standalone hub.

class PatchTest : public ::testing::Test {
 protected:
  void SetUp() override { Hub::Instance().InitOnce(); }

  std::shared_ptr<tableau::load::Options> NewOptions() const {
    auto options = std::make_shared<tableau::load::Options>();
    options->ignore_unknown_fields = true;
    return options;
  }
};

TEST_F(PatchTest, PatchConf_RecursivePatchConf_MatchesExpectedResult) {
  auto options = NewOptions();
  options->patch_dirs = {test::TestPaths::PatchConf().string()};

  bool ok = Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options);
  ASSERT_TRUE(ok) << "load failed: " << tableau::GetErrMsg();

  auto mgr = Hub::Instance().Get<protoconf::RecursivePatchConfMgr>();
  ASSERT_NE(mgr, nullptr);

  // Load expected golden result.
  tableau::RecursivePatchConf expected;
  ok = expected.Load(test::TestPaths::PatchResult().string() + "/", tableau::Format::kJSON);
  ASSERT_TRUE(ok) << "load expected failed: " << tableau::GetErrMsg();

  EXPECT_TRUE(MessageDifferencer::Equals(mgr->Data(), expected.Data()))
      << "actual: " << mgr->Data().ShortDebugString() << "\nexpected: " << expected.Data().ShortDebugString();
}

TEST_F(PatchTest, PatchConf_PatchReplaceConf_ReplacesEntirely) {
  auto options = NewOptions();
  options->patch_dirs = {test::TestPaths::PatchConf().string()};

  ASSERT_TRUE(Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options));
  auto mgr = Hub::Instance().Get<protoconf::PatchReplaceConfMgr>();
  ASSERT_NE(mgr, nullptr);
  EXPECT_EQ("orange", mgr->Data().name());
}

TEST_F(PatchTest, PatchConf2_LoadsSuccessfully) {
  auto options = NewOptions();
  options->patch_dirs = {test::TestPaths::PatchConf2().string()};
  ASSERT_TRUE(Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options));
}

TEST_F(PatchTest, PatchPaths_DifferentFormat_TxtPb) {
  auto options = NewOptions();
  options->patch_dirs = {test::TestPaths::PatchConf2().string()};
  auto mopts = std::make_shared<tableau::load::MessagerOptions>();
  mopts->patch_paths = {(test::TestPaths::PatchConf2() / "PatchMergeConf.txtpb").string()};
  options->messager_options["PatchMergeConf"] = mopts;

  bool ok = Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options);
  ASSERT_TRUE(ok) << "load failed: " << tableau::GetErrMsg();
  EXPECT_NE(Hub::Instance().Get<protoconf::PatchMergeConfMgr>(), nullptr);
}

TEST_F(PatchTest, PatchPaths_MultiplePatchFiles) {
  auto options = NewOptions();
  auto mopts = std::make_shared<tableau::load::MessagerOptions>();
  mopts->patch_paths = {(test::TestPaths::PatchConf() / "PatchMergeConf.json").string(),
                        (test::TestPaths::PatchConf2() / "PatchMergeConf.json").string()};
  options->messager_options["PatchMergeConf"] = mopts;

  ASSERT_TRUE(Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options));
  auto mgr = Hub::Instance().Get<protoconf::PatchMergeConfMgr>();
  ASSERT_NE(mgr, nullptr);
  // patchconf2's patch contributes key 999 into both ItemMap and ReplaceItemMap.
  EXPECT_TRUE(mgr->Data().item_map().contains(999)) << "ItemMap should contain key 999 from patchconf2";
  EXPECT_TRUE(mgr->Data().replace_item_map().contains(999)) << "ReplaceItemMap should contain key 999";
}

TEST_F(PatchTest, ModeOnlyMain_IgnoresPatches) {
  auto options = NewOptions();
  auto mopts = std::make_shared<tableau::load::MessagerOptions>();
  mopts->patch_paths = {(test::TestPaths::PatchConf() / "PatchMergeConf.json").string(),
                        (test::TestPaths::PatchConf2() / "PatchMergeConf.json").string()};
  options->messager_options["PatchMergeConf"] = mopts;
  options->mode = tableau::load::LoadMode::kOnlyMain;

  ASSERT_TRUE(Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options));
  auto mgr = Hub::Instance().Get<protoconf::PatchMergeConfMgr>();
  ASSERT_NE(mgr, nullptr);

  // Should equal a fresh OnlyMain load of the same file.
  tableau::PatchMergeConf direct;
  auto direct_opts = std::make_shared<tableau::load::MessagerOptions>();
  direct_opts->mode = tableau::load::LoadMode::kOnlyMain;
  ASSERT_TRUE(direct.Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, direct_opts));

  EXPECT_TRUE(MessageDifferencer::Equals(mgr->Data(), direct.Data()));
}

TEST_F(PatchTest, ModeOnlyPatch_AppliesPatchesFromEmpty) {
  auto options = NewOptions();
  auto mopts = std::make_shared<tableau::load::MessagerOptions>();
  mopts->patch_paths = {(test::TestPaths::PatchConf() / "PatchMergeConf.json").string(),
                        (test::TestPaths::PatchConf2() / "PatchMergeConf.json").string()};
  options->messager_options["PatchMergeConf"] = mopts;
  options->mode = tableau::load::LoadMode::kOnlyPatch;

  ASSERT_TRUE(Hub::Instance().Load(test::TestPaths::Conf().string() + "/", tableau::Format::kJSON, options));
  auto mgr = Hub::Instance().Get<protoconf::PatchMergeConfMgr>();
  ASSERT_NE(mgr, nullptr);
  // OnlyPatch starts from an empty message; Name must come from a patch file.
  EXPECT_FALSE(mgr->Data().name().empty());
}

}  // namespace
